File: 051_N-Queens/051_1.py
from typing import List

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        self.res = [0 for _ in range(n)]
        self.result = []
        self.solve(0, n)
        return self.result

    def solve(self, row, n):
        if (row == n):
            self.printQueen(n)
            return
        for col in range(n):
            if self.isValid(row, col, n):
                self.res[row] = col
                self.solve(row + 1, n)

    def isValid(self, row, col, n):
        leftUp, rightUp = col - 1, col + 1
        for i in range(row - 1, -1, -1):
            if self.res[i] == col:
                return False
            if leftUp >= 0:
                if self.res[i] == leftUp:
                    return False
            if rightUp < n:
                if self.res[i] == rightUp:
                    return False
            leftUp  -= 1
            rightUp += 1
        return True

    def printQueen(self, n):
        res = []
        for i in range(n):
            s = ""
            for j in range(n):
                if self.res[i] == j:
                    print("Q", end="")
                    s += "Q"
                else:
                    print(".", end="")
                    s += "."
            res.append(s)
            print()
        print()
        self.result.append(res)


if __name__ == "__main__":
    s = Solution()
    res = s.solveNQueens(4)
    print(res)

File: 051_N-Queens/051_1.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.*;

class Solution {
    List<List<String>> res = new ArrayList<>();
    public List<List<String>> solveNQueens(int n) {
        char[][] c = new char[n][n];
        for (int i = 0; i < n; i++)
            Arrays.fill(c[i], '.');
    //    List<char[]> one = new ArrayList<>();
    //    for (int i = 0; i < n; ++i)
    //        one.add(c);
        backtrack(c, 0);
        return res;
    }

    private void backtrack(char[][] one, int row) {
        int n = one.length;
        if (row == n) {
            addResult(one);
            return;
        }
            
        for (int col = 0; col < n; col++) {
            if (!isValid(one, row, col)) continue;
            one[row][col] = 'Q';
            backtrack(one, row + 1);
            one[row][col] = '.';
        }
    }
/*
    private boolean isValid(char[][] one, int row, int col) {
        int n = one.length;
        for (int i = 0; i < row; ++i) {
            if (one[i][col] == 'Q') return false;
        }
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (one[i][j] == 'Q') return false;
        }
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (one[i][j] == 'Q') return false;
        }
        return true;
    }
*/
    private boolean isValid(char[][] one, int row, int col) {
        int n = one.length;
        int leftUp = col - 1, rightUp = col + 1;
        for (int i = row - 1; i >= 0; --i) {
            if (one[i][col] == 'Q') return false;
            if (leftUp >= 0) 
                if (one[i][leftUp] == 'Q') return false;
            if (rightUp < n)
                if (one[i][rightUp] == 'Q') return false;
            leftUp--;
            rightUp++;
        }
        return true;
    }

    private void addResult(char[][] one) {
        int n = one.length;
        List<String> s = new ArrayList<>();
        for (int i = 0; i < n; ++i)
            s.add(new String(one[i]));
        System.out.println(s);
        res.add(s);
    }
}

public class leetcode {
    public static void main(String[] args) {
        Solution s = new Solution();
        List<List<String>> res = s.solveNQueens(4);  
    }
}